rust-symcrypt/src/rsa/mod.rs

```diff
@@ -0,0 +1,716 @@
+//!
+//! // key_pair can now be used via pkcs1, pss, or oaep.
+//!
+//! // Get the size of the modulus, size of the public exponent, and size of the primes.
+//! let modulus_size = key_pair.get_size_of_modulus();
+//! let pub_exp_size = key_pair.get_size_of_public_exponent();
+//! let primes = key_pair.get_size_of_primes();
+//!
+//! // Export the key pair to a blob.
+//! let key_pair_blob = key_pair.export_key_pair_blob().unwrap();
+//! ```
+//!
+//! ## Setting a RsaPublicKey object from parameters.
+//!
+//! ```rust
+//! use symcrypt::rsa::{RsaPublicKey, RsaKeyUsage};
+//!
+//! // Set an RsaPublicKey based on the modulus, public exponent, and key usage.
+//! let modulus = [
+//!     215, 145, 16, 194, 78, 246, 213, 23, 173, 178, 123, 179, 152, 238, 67, 16, 25, 20, 102,
+//!     36, 142, 210, 5, 164, 214, 122, 56, 206, 61, 65, 121, 44, 248, 241, 176, 72, 104, 251,
+//!     188, 59, 107, 251, 214, 238, 237, 49, 27, 224, 96, 114, 82, 54, 116, 238, 151, 56, 73,
+//!     216, 107, 88, 226, 27, 176, 247, 180, 48, 165, 127, 156, 133, 148, 69, 67, 191, 196,
+//!     148, 115, 123, 86, 185, 169, 42, 111, 109, 121, 21, 9, 174, 183, 126, 219, 3, 32, 83,
+//!     183, 63, 98, 253, 243, 108, 26, 9, 75, 68, 33, 248, 71, 223, 51, 231, 153, 194, 233,
+//!     245, 53, 86, 243, 164, 94, 123, 146, 75, 161, 99, 5, 145, 85, 165, 187, 146, 243, 196,
+//!     181, 223, 232, 32, 247, 253, 217, 211, 170, 187, 32, 23, 177, 11, 241, 133, 141, 8, 38,
+//!     124, 133, 88, 81, 230, 110, 200, 219, 28, 149, 77, 25, 163, 18, 75, 183, 210, 68, 0,
+//!     126, 3, 182, 196, 126, 207, 27, 92, 144, 174, 178, 203, 200, 146, 45, 180, 202, 3, 76,
+//!     22, 202, 37, 87, 215, 183, 83, 159, 65, 144, 9, 172, 137, 75, 17, 51, 31, 176, 6, 168,
+//!     197, 156, 195, 253, 36, 71, 64, 125, 253, 126, 155, 169, 79, 180, 233, 157, 193, 100,
+//!     239, 237, 129, 4, 165, 38, 112, 247, 253, 174, 21, 245, 71, 236, 229, 56, 123, 134, 45,
+//!     17, 124, 191, 60, 163, 218, 149, 209, 207, 181,
+//! ];
+//!
+//! let pub_exp = [0, 0, 0, 0, 0, 1, 0, 1];
+//! let public_key = RsaPublicKey::set_public_key(&modulus, &pub_exp, RsaKeyUsage::Sign).unwrap();
+//!    
+//! // public_key can now be used via pkcs1, pss, or oaep.
+//!
+//! // get the size of the modulus and the size of the public exponent.
+//! let modulus_size = public_key.get_size_of_modulus();
+//! let pub_exp_size = public_key.get_size_of_public_exponent();
+//!
+//! ```
+//!
+use crate::errors::SymCryptError;
+use crate::NumberFormat;
+use std::ptr;
+
+pub mod oaep;
+pub mod pkcs1;
+pub mod pss;
+
+/// !Review: Can we revisit RsaKeyUsage? It seems like it is not really needed since we dont actually do anything under the covers for it. 
+/// @ Mitch are the flags enforced at all by SymCrypt? If it's not, i think it might make it much easier to set symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_ENCRYPT | symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_SIGN
+/// We can enforce the usage in this layer but i dont feel good straying that far from symcrypt.
```

Comment:
The flags are enforced by SymCrypt; you can only encrypt/decrypt or sign/verify with a key that has the corresponding flag set. We should continue to let SymCrypt enforce this, but callers still need to be able to specify their key usage. It is generally bad practice to use a single key for both signing and encryption, although sometimes it's necessary, as in the case of cert enrollment. For some algorithms, the usage also determines which FIPS selftests are run, although I don't believe we currently do that for RSA. But that is an implementation detail that could change in the future.

Reply:
Khang, could you add tests to ensure KeyUsage is enforced by SymCrypt.

Reply:
Added